core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala

```diff
@@ -71,6 +82,159 @@ private[spark] class CoarseGrainedExecutorBackend(
+          tryConvertUnitsToBytes(userExecConfigCount, userExecConfigUnits.get)
+        } else {
+          userExecConfigCount.toLong
+        }
+        if (execConfigCountWithUnits != realCounts.execCount) {
+          throw new SparkException(s"Executor resource: $rName, count: ${realCounts.execCount} " +
+            s"doesn't match what user requests for executor count: $execConfigCountWithUnits, " +
+            s"via $userExecCountConfigName")
+        }
+      } else {
+        throw new SparkException(s"Executor resource config missing required task resource: $rName")
+      }
+    }
+  }
+
+  // visible for testing
+  def parseResources(resourcesFile: Option[String]): Map[String, ResourceInformation] = {
+    // only parse the resources if a task requires them
+    val taskResourceConfigs = env.conf.getAllWithPrefix(SPARK_TASK_RESOURCE_PREFIX)
+    val resourceInfo = if (taskResourceConfigs.nonEmpty) {
+      val execResources = resourcesFile.map { resourceFileStr => {
+        val source = new BufferedInputStream(new FileInputStream(resourceFileStr))
+        val resourceMap = try {
+          val parsedJson = parse(source).asInstanceOf[JArray].arr
+          parsedJson.map(_.extract[ResourceInformation]).map(x => (x.name -> x)).toMap
+        } catch {
+          case e @ (_: MappingException | _: MismatchedInputException | _: ClassCastException) =>
+            throw new SparkException(
+              s"Exception parsing the resources in $resourceFileStr", e)
+        } finally {
+          source.close()
+        }
+        resourceMap
+      }}.getOrElse(ResourceDiscoverer.findResources(env.conf, false))
+
+      if (execResources.isEmpty) {
+        throw new SparkException(s"User specified resources per task via: " +
+          s"$SPARK_TASK_RESOURCE_PREFIX, but can't find any resources available on the executor.")
+      }
+
+      // check that the executor has all the resources required by the application/task
+      checkExecResourcesMeetTaskRequirements(taskResourceConfigs, execResources)
+
+      // make sure the addresses make sense with count
+      execResources.foreach { case (rName, rInfo) =>
+        // check to make sure we have enough addresses when any specified, if we have
+        // more don't worry about it
+        if (rInfo.addresses.nonEmpty && rInfo.addresses.size < rInfo.count) {
+          throw new SparkException(s"The number of resource addresses is expected to either " +
+            s"be >= to the count or be empty if not applicable! Resource: $rName, " +
+            s"count: ${rInfo.count}, number of addresses: ${rInfo.addresses.size}")
+        }
+      }
+
+      logInfo(s"Executor ${executorId} using resources: ${execResources.keys}")
+      if (log.isDebugEnabled) {
+        logDebug("===============================================================================")
+        logDebug("Executor Resources:")
+        execResources.foreach{ case (k, v) =>
```

Comment:
minor: simpler if we define `ResourceInformation.toString()`

Reply:
done